packages/jest-runtime/src/__tests__/runtime_compile_module.test.js

```diff
@@ -0,0 +1,79 @@
+/**
+ * Copyright (c) Facebook, Inc. and its affiliates. All Rights Reserved.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ *
+ */
+
+'use strict';
+
+import {Module, builtinModules} from 'module';
+import path from 'path';
+import {pathToFileURL} from 'url';
+// eslint-disable-next-line import/default
+import slash from 'slash';
+import * as fs from 'graceful-fs';
+
+let createRuntime;
+
+describe('Runtime requireModule', () => {
+  beforeEach(() => {
+    createRuntime = require('createRuntime');
+  });
+
+  it('overrides `Module#compile`', () =>
+    createRuntime(__filename).then(runtime => {
+      const exports = runtime.requireModule(runtime.__mockRootPath, 'module');
+      expect(exports.Module).not.toBe(Module);
+
+      {
+        const pathRegularModule = path.join(
+          path.dirname(runtime.__mockRootPath),
+          'RegularModule.js',
+        );
+        const source = fs.readFileSync(pathRegularModule, 'utf-8');
+
+        const module = new exports.Module();
+        module._compile(source, pathRegularModule);
+        expect(module).toMatchObject({
+          children: expect.anything(),
+          exports: expect.anything(),
+          filename: expect.anything(),
+          id: expect.anything(),
+          loaded: false,
+          parent: null,
+          paths: expect.anything(),
+        });
+        expect(Object.keys(module.exports)).toEqual([
+          'filename',
+          'getModuleStateValue',
+          'isRealModule',
+          'jest',
+          'lazyRequire',
+          'object',
+          'parent',
+          'paths',
+          'setModuleStateValue',
+          'module',
+          'loaded',
+          'isLoaded',
+        ]);
+
+        expect(module.exports.getModuleStateValue()).toBe('default');
+
+        module.exports.lazyRequire();
+
+        expect(module.exports.getModuleStateValue()).toBe('Side effect value');
+      }
+
+      {
+        const module = new exports.Module();
+        module._compile(
+          'exports.value = 12;',
+          `${runtime.__mockRootPath}/dynamic.js`,
+        );
+        expect(module.exports).toEqual({value: 12});
```

Comment:
Does this actually verify that the module was run in the runtime, with access to the environment etc.?

Reply:
It really just exists to verify that it doesn't break basic functionality. The other test _does_ verify that the module was run in the runtime because it verifies that `jest` is available. I do intend to add another test that checks that jsdom is present, just as soon as I figure out how to do that in this test setup.

Reply:
b1c49a328